RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,208 @@
+    std::vector<T>& V_sk, // length at least d*n 
+    std::vector<T>& sigma_sk, // length at least n
+    int64_t threads, // number of OpenMP threads to use in sketching.
+    uint64_t seed_key,
+    uint32_t seed_ctr
+) {
+    T* buff_A = A.data();
+    if (V_sk.size() < d*n)
+        throw std::invalid_argument("V_sk must be of size at least d*n.");
+    T* buff_A_sk = V_sk.data();
+    
+    // step 1.1: define the sketching operator
+    RandBLAS::sparse::SparseDist D{
+        .n_rows = d,
+        .n_cols = m,
+        .vec_nnz=k
+    };
+    auto state = RandBLAS::base::RNGState(seed_key, seed_ctr);
+    RandBLAS::sparse::SparseSkOp<T> S(D, state);
+    RandBLAS::sparse::fill_sparse(S);
+
+    // step 1.2: sketch the data matrix
+    int64_t lda, lda_sk;
+    if (layout_A == blas::Layout::RowMajor) {
+        lda = n;
+        lda_sk = n;
+    } else {
+        lda = m;
+        lda_sk = d;
+    }
+    blas::scal(d*n, 0.0, buff_A_sk, 1);
+    RandBLAS::sparse::lskges<T>(
+        layout_A,
+        blas::Op::NoTrans,
+        blas::Op::NoTrans,
+        d, n, m,
+        1.0, S, 0, 0,
+        buff_A, lda,
+        0.0, buff_A_sk, lda_sk
+    );
+
+    // step 2: compute SVD of sketch
+    T* ignore;
+    if (layout_A == blas::Layout::RowMajor) {
+        //
+        //      buff_A_sk is stored in row-major format; we want its singular values
+        //      and right singular vectors.
+        //
+        //      We get around LAPACK's column-major restriction by using the fact that
+        //          memory(A_sk, RowMajor) == memory(transpose(A_sk), ColumnMajor).
+        //      So we'll tell LAPACK that we want the singular values and left
+        //      singular vectors of tranpose(A_sk).
+        //
+        //      TODO: check that rank(A_sk) == rank(A). This can be done by 
+        //      taking a basis "B" for the kernel of A_sk and checking if
+        //      if A*B == 0 (up to some tolerance).   
+        //
+        lapack::Job jobu = lapack::Job::OverwriteVec;
+        lapack::Job jobvt = lapack::Job::NoVec;
+        lapack::gesvd(jobu, jobvt, n, d, buff_A_sk, n, sigma_sk.data(), ignore, 1, ignore, 1);
+        // interpret buff_A_sk in row-major
+    } else {
+        lapack::Job jobu = lapack::Job::NoVec;
+        lapack::Job jobvt = lapack::Job::SomeVec;
+        T *Vt = new T[n * n];
+        lapack::gesvd(jobu, jobvt, d, n, buff_A_sk, d, sigma_sk.data(), ignore, 1, Vt, n);
+        for (int i = 0; i < n; ++i) {
+            // buff_A_sk is now just scratch space owned by the std::vector V_sk.
+            // We just computed its SVD as a d-by-n matrix in column-major order.
+            // Now we want to write to the first n^2 elements of buff_A_sk with
+            // n "n-vectors" in column-major order.
+            blas::copy(n, &Vt[i], n,  &buff_A_sk[i*n], 1);
+        }
+        delete[] Vt;
+    }
+    return;
+}
+
+template <typename T>
+int64_t rpc_svd_saso(
```

Comment:
No docs?

Reply:
Not yet, lol. Will add before merging.

Reply:
This function has been removed. Accessing equivalent functionality now requires two function calls. The first call is to ``rpc_data_svd_saso``, and the second call is to ``make_rpc_svd_explicit``. The latter of these functions is deterministic, its name is subject to change, and documentation is forthcoming.